openfoodfoundation/openfoodnetwork
# Add new Stripe payment method compatible with the new Stripe Payment Intents API

#### What? Why?

Relates to #4180
Closes #4686 and #4173

The decision to duplicate the existing Stripe payment method has the objective of making the migration as smooth as possible and also of making it possible for us, after we migrate all users, to delete the original stripe payment method (and its code), and keep only the new one (that uses the new SCA compatible APIs).
Although this new API was made by stripe for European SCA it does work worldwide, we don't need to keep both integrations :+1:

This PR is not introducing any new auth processes yet. It should just work like the existing stripe payment method.

I have added entries to rubocop_manual_todo for the remaining rubocop issues. Code climate is not picking the exceptions list from rubocop_manual_todo...

In terms of tests we are not adding any feature tests here as we plan to get something new with #4639

#### What should we test?
We need to verify that the existing Stripe payment method works.
We need to verify the new payment method, it should just work like the existing stripe payment method.
Processes we need to test:
1. make a payment without saving card
2. make a payment saving card
3. make a payment using previously saved card
4. add a card on account section and use it to checkout
5. delete saved card
6. capture the payment in the admin payments section of an order
7. as a user authorize a hub to use your card and verify the subscription payments work with that card

The following details are valid before #4719, if we are testing 4719, all cases above should work.
<details>
As discussed in comments below: case 4 above is not working yet for this payment method. But case 7, subscriptions, is working.
Because case 4 doesnt work (if you add a card in the account section and use it in a subscription with this new payment methods, the payments will fail). To add a card that works for a subscription with this new payment method you need to checkout an order with this new payment method with the save card option enabled. After that go to account to authorize the card you have used to checkout.
</details>

#### Release notes
Changelog Category: Added
Added new payment method that integrates with the new Stripe Payment Intents API that is compatible with SCA. This is a initial preparation for us to implement the more complex payment authentication flows that banks will start to enforce in Europe soon.
